content/en/blog/_posts/2024-09-17-devsecops-in-action-k8s-secrets/index.md

```diff
@@ -0,0 +1,554 @@
+development process and apply to all other levels correspondingly.
+
+According to
+[OWASP Secrets Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#5-containers-orchestrators),
+it's recommended to use Injection of Secret via file and in-memory. Provision of
+sensitive data via environment variables isn't recommended, but at the same time
+it's recommended in The
+[Twelve Factors](https://12factor.net/config).
+These are general recommendations which don't impact security issues that's why
+their application should make sense, but passing config via environment variables
+with sensitive data must be kept to a minimum.
+
+Injection of Secret via file and in-memory has pros and cons and I would like to
+show which security issues exist when using Injection of Secret via file and
+suggest solutions. In fact, there will be a general solution and specific
+implementation will be different in terms of specific management and delivery
+secrets tools. To keep the examples simple I will use `spec.initContainers` to
+simulate an injection agent for delivering secrets to Pod.
+
+## Baseline application
+
+Imagine there is a simple application which processes http requests. On request
+to `root` location client receives environment variable value
+`DEMO_SECRET__PASSWD` which is defined via
+[`spec.containers.0.envFrom.secretRef.name`](kustomization/patches/baseline.yaml)
+in Pod resource. On request `/readiness` receives status application
+(ready or not ready) and if the environment variable isn't defined Pod doesn't
+change status to `READY`.
+
+[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/14624762f1546e2bfd43e65cee26186c514fa351/cmd/main.go#L10-L53)
+```go
+const (
+	envVar = "DEMO_SECRET__PASSWD"
+)
+
+...
+
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+
+	var isExist, ok bool
+	var envVal string
+
+	if envVal, ok = os.LookupEnv(envVar); ok {
+		isExist = true
+	}
+
+	switch r.URL.Path {
+	case "/":
+		if isExist {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, envVal)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Secret Not Found")
+		}
+	case "/readiness":
+		if isExist {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "OK")
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Environment variable wasn't defined")
+		}
+	default:
+		w.WriteHeader(http.StatusNotFound)
+		fmt.Fprintln(w, "404 Not Found")
+	}
+}
+```
+
+## Stage-01: Reject environment vars for storing sensitive data
+
+According to OWASP recommendation, it's necessary to use file instead of env.
+That's why it's necessary to make changes to implementation readiness probes
+and deployment manifests.
+
+[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/96afd59f8a391973a8ed5a9b410915b9e99d4bf4/cmd/main.go#L11-L67)
```

Comment:
We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Reply:
Fixed